javacutil/src/main/java/org/checkerframework/javacutil/SwitchExpressionScanner.java

```diff
@@ -0,0 +1,171 @@
+import java.util.function.BiFunction;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * A class that visits each result expression of a switch expression and calls {@link
+ * #visitSwitchResultExpression(ExpressionTree, Object)} on each result expression. The results of
+ * these method calls are combined using {@link #combineResults(Object, Object)}. Call {@link
+ * #scanSwitchExpression(Tree, Object)} to start scanning the switch expression.
+ *
+ * <p>{@link FunctionalSwitchExpressionScanner} can be used to pass functions for to use for {@link
+ * #visitSwitchResultExpression(ExpressionTree, Object)} and {@link #combineResults(Object,
+ * Object)}.
+ *
+ * @param <R> the type of the result of {@link #visitSwitchResultExpression(ExpressionTree, Object)}
+ * @param <P> the type of the parameter to pass to {@link
+ *     #visitSwitchResultExpression(ExpressionTree, Object)}
+ */
+public abstract class SwitchExpressionScanner<R, P> extends TreeScanner<R, P> {
+
+  /**
+   * This method is called for each result expression of the switch expression passed in {@link
+   * #scanSwitchExpression(Tree, Object)}.
+   *
+   * @param resultExpressionTree a result expression of the switch expression currently being
+   *     scanned
+   * @param p a parameter
+   * @return the result of visiting the result expression
+   */
+  protected abstract R visitSwitchResultExpression(ExpressionTree resultExpressionTree, P p);
+
+  /**
+   * This method combines the result of two calls to {@link
+   * #visitSwitchResultExpression(ExpressionTree, Object)} or {@code null} and the result of one
+   * call to {@link #visitSwitchResultExpression(ExpressionTree, Object)}.
+   *
+   * @param r1 a possibly null result returned by {@link
+   *     #visitSwitchResultExpression(ExpressionTree, Object)}
+   * @param r2 a possibly null result returned by {@link
+   *     #visitSwitchResultExpression(ExpressionTree, Object)}
+   * @return the combination of {@code r1} and {@code r2}
+   */
+  protected abstract R combineResults(@Nullable R r1, @Nullable R r2);
+
+  /**
+   * Scans the given switch expression and calls {@link #visitSwitchResultExpression(ExpressionTree,
+   * Object)} on each result expression of the switch expression. {@link #combineResults(Object,
+   * Object)} is called to combine the results of visiting multiple switch result expressions.
+   *
+   * @param switchExpression a switch expression tree
+   * @param p the parameter to pass to {@link #visitSwitchResultExpression(ExpressionTree, Object)}
+   * @return the result of calling {@link #visitSwitchResultExpression(ExpressionTree, Object)} on
+   *     each result expression of {@code switchExpression} and combining the results using {@link
+   *     #combineResults(Object, Object)}
+   */
+  public R scanSwitchExpression(Tree switchExpression, P p) {
+    assert switchExpression.getKind().name().equals("SWITCH_EXPRESSION");
```

Comment:
Currently, we never use Java 17 to compile any part of the Checker Framework.  We could, but it would introduce complexity in the build and release scripts.